### DISCUSSION TITLE
for signup with confirmation method, users with unconfirmed emails can signup again with the same emails but different passwords

### DISCUSSION THREAD

**Discussion by devhandler** [supabase/supabase#14994] [Feature Requests]
but in db, it's actually still the same password when the users first signed up. 

now after users confirm their emails this time, they won't be able to log in because they think the passwords are what they just submitted instead of the one during their first signups.  this could be very confusing to users. i'm wondering if anyone has some feasible solutions? 

ofc, what's really needed is the auth table can actually update the password whenever users use the same email (but unconfirmed) to signup again.

**Comment by davidguy3237** [comment]
I'm running into the same issue. Would like to know if anyone has found a solution or a workaround.

**Reply by devhandler** [reply]
one workaround could be checking the time difference between created at and updated at and throw error to users. so users know this time they didn't sign up successfully and wouldn't use the password they type this time.

**Reply by davidguy3237** [reply]
I think I found a workaround that seems to work, although I am unsure if there are any potential security issues in doing it this way. If you decide to go this route, you're doing so at your own risk. I'm still quite inexperienced when it comes to stuff like this, so any feedback would be much appreciated.

Lurking around, I found #4042 which talks about confirming your password before being able to change it. The solution recommends [invoking a Postgres function via a Remote Procedure Call (RPC)](https://supabase.com/docs/reference/javascript/rpc) which will check if the old password is the same before updating it.
I figured I could kind of use the same logic where after signing up, it'll check if the account exists, if the email is still unverified, and if it is, it'll replace the existing password in the database.

Here is my Postgres function:
```
CREATE OR REPLACE FUNCTION update_unverified_account_password (user_id uuid, password text)
RETURNS json
LANGUAGE plpgsql
SECURITY DEFINER
AS
$$
DECLARE
uid uuid;
BEGIN
  -- Verify that submitted password conforms to rules
  IF (password = '') IS TRUE THEN
    RAISE EXCEPTION 'Submitted password is empty';
  ELSEIF char_length(password) < 8 THEN
    RAISE EXCEPTION 'Password must be at least 8 characters';
  END IF;

  -- Check if account exists, if the email has been verified, and if the submitted password is different than the password in the database.
  SELECT id INTO uid
  FROM auth.users
  WHERE id = user_id
  AND email_confirmed_at IS NULL
  AND encrypted_password != crypt(password::text, auth.users.encrypted_password);

  IF NOT FOUND THEN
    RAISE EXCEPTION 'PASSWORD MATCHES OR EMAIL IS CONFIRMED';
  END IF;

  -- Replace password in database with the one that was passed in.
  UPDATE auth.users SET
  encrypted_password = crypt(password::text, gen_salt('bf'))
  WHERE id = user_id;

  RETURN '{"data":"success"}';
END;
$$;
```
Since [auth.uid\(\) only works if you're signed in](https://supabase.com/docs/guides/auth#:~:text=Supabase%20provides%20a%20special%20function%20in%20Postgres%2C%20auth.uid\(\)%2C%20which%20extracts%20the%20user%27s%20UID%20from%20the%20JWT.%20This%20is%20especially%20useful%20when%20creating%20policies.), I had to opt for manually passing in the uuid into the function, which I got from the data object that gets returned by `supabase.auth.signUp(...)`. Not sure if there are any security concerns doing it this way, please let me know if there is. I suppose it would also work if I used email instead of the uuid.

For my server logic, I haven't completely fleshed it out, but here's the basic gist:

Upon submitting the sign up form, invoke `supabase.auth.signUp()`

if there's an error, then handle the logic for that

else if `data.user.identities.length === 0` then email is already taken and verified

else if `data.user.identities.length !== 0` then email is not taken or at least not verified, so then invoke `supabase.rpc("update_unverified_account_password", { user_id, password})` which will replace the password in the database if the email is still unverified.
